core/include/kernel/interrupt.h

```diff
@@ -12,13 +12,26 @@
 #define ITRF_TRIGGER_LEVEL	BIT(0)
 #define ITRF_SHARED			BIT(1)
 
+#define IRQ_MAX_SPEC		4
+
+/* struct itr_spec - Interrupt specifier
+ * @irq_spec: Specifier for interrupt
+ * @num_spec: number of specifiers in the irq_spec member
+ */
+struct itr_spec {
+	uint32_t irq_spec[IRQ_MAX_SPEC];
```

Comment:
This makes `struct itr_handler` to grow even for platforms when useless (not that there are many that care). Can the struct be allocated to the expected size when needed?

I don't really like such opaque an blob list. Things in _interrupt.h_ will be generic. It looks like this is expected to store a raw 32bit data series from the DT and pass to back-end driver. If data are driver specific then the driver should allocate a private data area in the interrupt instance and store there any driver specific information.

Reply:
Agreed, the idea was to be able to pass interrupt description (often described in the devicetree) to the driver. 
Currently, dt_get_irq returns only an int property, and even if not used, how do you pass additional information to the interrupt driver at itr_add call ? flags members seems a bit limiting.

Reply:
To be more clear, I need to pass priority and interrupt type (LEVEL, EDGE, HIGH, LOW) to the interrupt controller. This information is described in the device tree but might be provided by non dt code. These data are usually described using the interrupt cells (which are uint32_t).

One possibility is probably to drop this from struct itr_spec and pass it to itr_add and others. This way it won't take any space at all (except on the stack when declaring it). But it will be decoupled from the itr_handler which seems a bit odd.

Reply:
> But it will be decoupled from the itr_handler which seems a bit odd.

If `itr_add()` blindly passes it as an arg to the controller function handler (`struct itr_ops::add`), it looks fine to me. These are controller specific config data, so fine if opaque to `struct itr_handler`, It could even replace the `flags` argument in `itr_add()` that serves an equivalent purpose and is currently unused by the sole interrupt controller supported in optee.

Reply:
For dt_fill_device_info, the controller will add to allocate data because this part is generic and call dt_get_irq.
This means that dt_get_irq  will return an allocated struct by the controller. This is probably adding unnecessary allocations, and IMHO, the interrupt specifier that are used in the dt are simple (up to 4 uint32_t) and seems to fulfill their purpose to describe interrupts for almost all interrupt controllers.

Reply:
> This is probably adding unnecessary allocations 

These data are used only during `itr_add()` for configuration, right? I think they could be de-referenced/released once `itr_add()` completes.

> IMHO, the interrupt specifier that are used in the dt are simple (up to 4 uint32_t) and seems to fulfill their purpose to describe interrupts for almost all interrupt controllers.

The interrupt controller driver could read itself into the DT to get/process expected DT property cells. Maybe the DT parsing should not be done in generic `interrupt.c`.  Maybe over killing :).
Note the optee implements a single interrupt controller. If we need to support more, possibly cascade controllers, _interrupt.c|.h_ current API will not match.

As for `dt_fill_device_info()`, it is a convenient helper function and could remain simple IMO and always call `dt_get_irq()` with a NULL reference for extra interrupt configs.

I think we could remove field `struct itr_handler::flags` and replace existing argument `flags` of `itr_add()` with this `struct itr_dt_spec` reference, which could possibly be NULL.

Reply:
> > This is probably adding unnecessary allocations
> 
> These data are used only during `itr_add()` for configuration, right? I think they could be de-referenced/released once `itr_add()` completes.

> 
> > IMHO, the interrupt specifier that are used in the dt are simple (up to 4 uint32_t) and seems to fulfill their purpose to describe interrupts for almost all interrupt controllers.
> 
> The interrupt controller driver could read itself into the DT to get/process expected DT property cells. Maybe the DT parsing should not be done in generic `interrupt.c`. Maybe over killing :).

As I said, this might be done without devicetree, it is still necessary to passe information to the interrupt controller.

> Note the optee implements a single interrupt controller. If we need to support more, possibly cascade controllers, _interrupt.c|.h_ current API will not match.

The controller I've added does not requires cascaded interrupts but only needs more attributes for irq configuration so the current API is ok.

> 
> As for `dt_fill_device_info()`, it is a convenient helper function and could remain simple IMO and always call `dt_get_irq()` with a NULL reference for extra interrupt configs.

But then, device tree interrupts property won't be parsed and won't be retrieval by the drivers. This seems a bit deceiving.

> 
> I think we could remove field `struct itr_handler::flags` and replace existing argument `flags` of `itr_add()` with this `struct itr_dt_spec` reference, which could possibly be NULL.

I guess you talk about `itr_ops::add` since itr_add only takes a handler as parameter.

Reply:
> > As for `dt_fill_device_info(),` it is a convenient helper function and could remain simple IMO and always call `dt_get_irq()` with a NULL reference for extra interrupt configs.
> 
> But then, device tree interrupts property won't be parsed and won't be retrieval by the drivers. This seems a bit deceiving.

Sorry, i meant "(...) remain simple IMO and **could** always call `dt_get_irq()` with a NULL (...)".
I have no objection for `dt_fill_device_info()` to povide more help.


> > I think we could remove field `struct itr_handler::flags` and replace existing argument `flags` of `itr_add()` with this `struct itr_dt_spec` reference, which could possibly be NULL.
> 
> I guess you talk about `itr_ops::add` since itr_add only takes a handler as parameter.

Right. Add this `_spec` reference as argument to `itr_add()` and replace `itr_ops::add()` argument `flags` with the `_spec` reference.

Reply:
Ok, actually after looknig a bit more carefully at it, `flags` is used for more generic attributes and is used by the generic interrupt code:
```
			assert((hdl->flags & ITRF_SHARED) &&
			       (h->flags & ITRF_SHARED));
```
I'm not sure it would be a good idea to remove it. The struct `itr_spec` is for hardware interrupt line description and is meant for the interrupt controller driver.

Reply:
Me to blame, you're right. `flags` is internal. It already provides the edge/level config but is generic in `itr_` framework.

Reply:
Ok, so I will pass NULL when calling `dt_get_irq` in `dt_fill_device_info`. If needed, it could be added later.
I will also remove the flag argument for `.add` callback since flags should remain generic for the itr framework and keep the itr_spec for this callback which will contain the expected interrupt description (same as in the device tree) for the interrupt controller driver .